target/arc/semfunc-v2.c

```diff
@@ -7260,159 +7227,106 @@ arc_gen_ST(DisasCtxt *ctx, TCGv src1, TCGv src2, TCGv dest)
-        getBit(temp_6, dest, temp_7);
-        tcg_gen_setcondi_tl(TCG_COND_EQ, temp_1, temp_6, 1);
-        tcg_gen_xori_tl(temp_2, temp_1, 1);
-        tcg_gen_andi_tl(temp_2, temp_2, 1);
-        tcg_gen_brcond_tl(TCG_COND_EQ, temp_2, arc_true, done_1);
-        tcg_gen_movi_tl(tmp, 4294967295);
-        gen_set_label(done_1);
-        tcg_gen_addi_tl(temp_8, address, 4);
-        setMemory(temp_8, LONG, tmp);
+        /* sign-extend to 64 bits if it's an immediate */
+        TCGv value_hi = tcg_temp_new();
+        tcg_gen_andi_tl(value_hi, dest, 1U << 31);
+        tcg_gen_movcond_tl(TCG_COND_NE, value_hi, value_hi, tcg_constant_tl(0),
+                        tcg_constant_tl(0xFFFFFFFF), tcg_constant_tl(0));
+
+        tcg_gen_addi_tl(address, address, 4);
+        tcg_gen_qemu_st_tl(value_hi, address, ctx->mem_idx, MO_UL);
+
+        tcg_temp_free(value_hi);
     }
-    if (((AA == 1) || (AA == 2))) {
+
+    /* update destination register if needed */
+    if (aa == 1 || aa == 2) {
         tcg_gen_add_tl(src1, src1, src2);
     }
+
+    tcg_temp_free(scaled_disp);
     tcg_temp_free(address);
-    tcg_temp_free(temp_3);
-    tcg_temp_free(temp_4);
-    tcg_temp_free(temp_5);
-    tcg_temp_free(tmp);
-    tcg_temp_free(temp_7);
-    tcg_temp_free(temp_6);
-    tcg_temp_free(temp_1);
-    tcg_temp_free(temp_2);
-    tcg_temp_free(temp_8);
 
-    return ret;
+    return DISAS_NEXT;
 }
 
 
 /*
  * POP
  *    Variables: @dest
- *    Functions: getMemory, getRegister, setRegister
+ *    Functions: getMemory, arc_gen_get_register, setRegister
  * --- code ---
  * {
- *   new_dest = getMemory (getRegister (R_SP), LONG);
- *   setRegister (R_SP, (getRegister (R_SP) + 4));
+ *   new_dest = getMemory (arc_gen_get_register (R_SP), LONG);
+ *   setRegister (R_SP, (arc_gen_get_register (R_SP) + 4));
  *   @dest = new_dest;
  * }
  */
 
 int
 arc_gen_POP(DisasCtxt *ctx, TCGv dest)
 {
-    int ret = DISAS_NEXT;
-    TCGv temp_3 = tcg_temp_local_new();
-    TCGv temp_2 = tcg_temp_local_new();
-    TCGv temp_1 = tcg_temp_local_new();
-    TCGv new_dest = tcg_temp_local_new();
-    TCGv temp_6 = tcg_temp_local_new();
-    TCGv temp_5 = tcg_temp_local_new();
-    TCGv temp_4 = tcg_temp_local_new();
-    getRegister(temp_3, R_SP);
-    tcg_gen_mov_tl(temp_2, temp_3);
-    getMemory(temp_1, temp_2, LONG);
-    tcg_gen_mov_tl(new_dest, temp_1);
-    getRegister(temp_6, R_SP);
-    tcg_gen_mov_tl(temp_5, temp_6);
-    tcg_gen_addi_tl(temp_4, temp_5, 4);
-    setRegister(R_SP, temp_4);
-    tcg_gen_mov_tl(dest, new_dest);
-    tcg_temp_free(temp_3);
-    tcg_temp_free(temp_2);
-    tcg_temp_free(temp_1);
-    tcg_temp_free(new_dest);
-    tcg_temp_free(temp_6);
-    tcg_temp_free(temp_5);
-    tcg_temp_free(temp_4);
+    tcg_gen_qemu_ld_tl(dest, cpu_sp, ctx->mem_idx, MO_UL);
+    if (dest != cpu_sp)
+	    tcg_gen_addi_tl(cpu_sp, cpu_sp, 4);
 
-    return ret;
+    return DISAS_NEXT;
 }
 
 
 /*
  * PUSH
  *    Variables: @src
```

Comment:
This comment is true for all the functions you rewrote in semfunc-v{2,3}.
Get rid of "Variables" and "Functions" section. Make the `--- code ---` section as such to reflect the status quo.

See [SETI](https://github.com/foss-for-synopsys-dwc-arc-processors/qemu/blob/master/target/arc/semfunc-v2.c#L3979) as an example.